From 30a1507e7fa55bc726ade5d90a08446272c021e4 Mon Sep 17 00:00:00 2001
From: Simon Zolin <stsaz@yandex.ru>
Date: Sun, 6 Sep 2015 15:34:36 +0300
Subject: [PATCH 2/3] - the last frame was not decoded because there is no next
 data of size MAD_BUFFER_GUARD

Add stream option MAD_OPTION_LASTDATA which is used to determine whether it's the last frame.
In this case layer3 code doesn't attempt to read the data from the next frame to determine md_len.
---
 frame.c  | 3 ++-
 layer3.c | 2 +-
 mad.h    | 2 ++
 stream.h | 2 ++
 4 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/frame.c b/frame.c
index f2ad1c4..9fd7c2f 100644
--- a/frame.c
+++ b/frame.c
@@ -406,7 +406,8 @@ int mad_header_decode(struct mad_header *header, struct mad_stream *stream)
   }
 
   /* verify there is enough data left in buffer to decode this frame */
-  if (N + MAD_BUFFER_GUARD > end - stream->this_frame) {
+  if (N + MAD_BUFFER_GUARD > end - stream->this_frame
+    && (!(stream->options & MAD_OPTION_LASTFRAME) || N > end - stream->this_frame)) {
     stream->next_frame = stream->this_frame;
 
     stream->error = MAD_ERROR_BUFLEN;
diff --git a/layer3.c b/layer3.c
index 4e5d3fa..ccce43c 100644
--- a/layer3.c
+++ b/layer3.c
@@ -2582,7 +2582,7 @@ int mad_layer_III(struct mad_stream *stream, struct mad_frame *frame)
 
   /* find main_data of next frame */
 
-  {
+  if (stream->next_frame != stream->bufend) {
     struct mad_bitptr peek;
     unsigned long header;
 
diff --git a/mad.h b/mad.h
index 9ef6cc8..961f5ad 100644
--- a/mad.h
+++ b/mad.h
@@ -725,6 +725,8 @@ enum {
   MAD_OPTION_RIGHTCHANNEL   = 0x0020,	/* decode right channel only */
   MAD_OPTION_SINGLECHANNEL  = 0x0030	/* combine channels */
 # endif
+  ,
+  MAD_OPTION_LASTFRAME = (1 << 31),
 };
 
 void mad_stream_init(struct mad_stream *);
diff --git a/stream.h b/stream.h
index 8bff8b4..95a0ae5 100644
--- a/stream.h
+++ b/stream.h
@@ -89,6 +89,8 @@ enum {
   MAD_OPTION_RIGHTCHANNEL   = 0x0020,	/* decode right channel only */
   MAD_OPTION_SINGLECHANNEL  = 0x0030	/* combine channels */
 # endif
+  ,
+  MAD_OPTION_LASTFRAME = (1 << 31),
 };
 
 void mad_stream_init(struct mad_stream *);
-- 
2.4.3

